src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -296,9 +296,18 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                     IndexMetaData indexMeta = clusterState.getMetaData().indices().get(shardId.getIndex());
                     try {
                         indicesService.deleteShardStore("no longer used", shardId, indexMeta);
-                    } catch (Exception ex) {
+                    } catch (Throwable ex) {
                         logger.debug("{} failed to delete unallocated shard, ignoring", ex, shardId);
                     }
+                    // if the index doesn't exists anymore, delete its store as well, but only if its a non master node, since master
+                    // nodes keep the index metadata around 
+                    if (indicesService.hasIndex(shardId.getIndex()) == false && currentState.nodes().localNode().masterNode() == false) {
```

Comment:
Just wondering why would master nodes want to keep the index directories around?